## ANZ Data@ANZ Program 

This is from www.theforage.com from the Data@ANZ program 

There are 5 tasks:
    
    -Exploratory Data Analysis (Mandatory)
    -Predictive Analytics (Mandatory)
    -Exploring Big Data
    -Cloud Engineering Core Skills
    -Data Engineering Core Skills

"We are looking to distill unique insights from the data that will enable us to take action. We want to know “what does all this data actually mean for us and our clients, and what should we do next?”


### Some subtasks to be completed
    - Load the transaction dataset below into an analysis tool of your choice (Excel, R, SAS, Tableau, or similar)

    - Start by doing some basic checks – are there any data issues? Does the data need to be cleaned?

    - Gather some interesting overall insights about the data. For example -- what is the average transaction amount? How many transactions do customers make each month, on average?

    - Segment the dataset by transaction date and time. Visualise transaction volume and spending over the course of an average day or week. Consider the effect of any outliers that may distort your analysis.

    - For a challenge – what insights can you draw from the location information provided in the dataset?

    - Put together 2-3 slides summarising your most interesting findings to ANZ management.

## Cleaning the data

In [101]:
# Importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats
import seaborn as sns
import statsmodels.api as sm
import statistics
import datetime as dt

In [102]:
# Loading the provided excel file
df = pd.read_excel('ANZ_synthesised_transaction_dataset.xlsx', header = 0)



### Loading the dataframe for cleaning

In [103]:
df.head()

status  card_present_flag bpay_biller_code         account currency  \
0  authorized                1.0              NaN  ACC-1598451071      AUD   
1  authorized                0.0              NaN  ACC-1598451071      AUD   
2  authorized                1.0              NaN  ACC-1222300524      AUD   
3  authorized                1.0              NaN  ACC-1037050564      AUD   
4  authorized                1.0              NaN  ACC-1598451071      AUD   

        long_lat txn_description                           merchant_id  \
0  153.41 -27.95             POS  81c48296-73be-44a7-befa-d053f48ce7cd   
1  153.41 -27.95       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e   
2  151.23 -33.94             POS  835c231d-8cdf-4e96-859d-e9d571760cf0   
3  153.10 -27.66       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673   
4  153.41 -27.95       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0   

   merchant_code first_name  balance       date gender  age merchant_suburb  \
0            NaN      Diana    35.39 2018-08-01      F   26         Ashmore   
1            NaN      Diana    21.20 2018-08-01      F   26          Sydney   
2            NaN    Michael     5.71 2018-08-01      M   38          Sydney   
3            NaN     Rhonda  2117.22 2018-08-01      F   40         Buderim   
4            NaN      Diana    17.95 2018-08-01      F   26   Mermaid Beach   

  merchant_state                    extraction  amount  \
0            QLD  2018-08-01T01:01:15.000+0000   16.25   
1            NSW  2018-08-01T01:13:45.000+0000   14.19   
2            NSW  2018-08-01T01:26:15.000+0000    6.42   
3            QLD  2018-08-01T01:38:45.000+0000   40.90   
4            QLD  2018-08-01T01:51:15.000+0000    3.25   

                     transaction_id    country     customer_id  \
0  a623070bfead4541a6b0fff8a09e706c  Australia  CUS-2487424745   
1  13270a2a902145da9db4c951e04b51b9  Australia  CUS-2487424745   
2  feb79e7ecd7048a5a36ec889d1a94270  Australia  CUS-2142601169   
3  2698170da3704fd981b15e64a006079e  Australia  CUS-1614226872   
4  329adf79878c4cf0aeb4188b4691c266  Australia  CUS-2487424745   

  merchant_long_lat movement  
0     153.38 -27.99    debit  
1     151.21 -33.87    debit  
2     151.21 -33.87    debit  
3     153.05 -26.68    debit  
4     153.44 -28.06    debit

In [104]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12043 entries, 0 to 12042
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   status             12043 non-null  object        
 1   card_present_flag  7717 non-null   float64       
 2   bpay_biller_code   885 non-null    object        
 3   account            12043 non-null  object        
 4   currency           12043 non-null  object        
 5   long_lat           12043 non-null  object        
 6   txn_description    12043 non-null  object        
 7   merchant_id        7717 non-null   object        
 8   merchant_code      883 non-null    float64       
 9   first_name         12043 non-null  object        
 10  balance            12043 non-null  float64       
 11  date               12043 non-null  datetime64[ns]
 12  gender             12043 non-null  object        
 13  age                12043 non-null  int64         
 14  mercha

### Converting data types 

In [105]:
# Changing data types float64 to float32
df[['card_present_flag', 'merchant_code', 'balance', 'amount']] = df[['card_present_flag', 'merchant_code', 'balance', 'amount'
                                                                     ]].astype('float32')

# changing int64 to int 8
df['age'] = df['age'].astype('int8')

In [106]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12043 entries, 0 to 12042
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   status             12043 non-null  object        
 1   card_present_flag  7717 non-null   float32       
 2   bpay_biller_code   885 non-null    object        
 3   account            12043 non-null  object        
 4   currency           12043 non-null  object        
 5   long_lat           12043 non-null  object        
 6   txn_description    12043 non-null  object        
 7   merchant_id        7717 non-null   object        
 8   merchant_code      883 non-null    float32       
 9   first_name         12043 non-null  object        
 10  balance            12043 non-null  float32       
 11  date               12043 non-null  datetime64[ns]
 12  gender             12043 non-null  object        
 13  age                12043 non-null  int8          
 14  mercha

In [107]:
# The extraction column is in datetime format but is an object datatype
# Changing the date type
df['extraction'] = df['extraction'].apply(pd.to_datetime)

In [108]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12043 entries, 0 to 12042
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype              
---  ------             --------------  -----              
 0   status             12043 non-null  object             
 1   card_present_flag  7717 non-null   float32            
 2   bpay_biller_code   885 non-null    object             
 3   account            12043 non-null  object             
 4   currency           12043 non-null  object             
 5   long_lat           12043 non-null  object             
 6   txn_description    12043 non-null  object             
 7   merchant_id        7717 non-null   object             
 8   merchant_code      883 non-null    float32            
 9   first_name         12043 non-null  object             
 10  balance            12043 non-null  float32            
 11  date               12043 non-null  datetime64[ns]     
 12  gender             12043 non-null  object     

In [109]:
df['extraction'] = pd.to_datetime(df.extraction).dt.tz_localize(None)

In [110]:
df.head()

status  card_present_flag bpay_biller_code         account currency  \
0  authorized                1.0              NaN  ACC-1598451071      AUD   
1  authorized                0.0              NaN  ACC-1598451071      AUD   
2  authorized                1.0              NaN  ACC-1222300524      AUD   
3  authorized                1.0              NaN  ACC-1037050564      AUD   
4  authorized                1.0              NaN  ACC-1598451071      AUD   

        long_lat txn_description                           merchant_id  \
0  153.41 -27.95             POS  81c48296-73be-44a7-befa-d053f48ce7cd   
1  153.41 -27.95       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e   
2  151.23 -33.94             POS  835c231d-8cdf-4e96-859d-e9d571760cf0   
3  153.10 -27.66       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673   
4  153.41 -27.95       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0   

   merchant_code first_name      balance       date gender  age  \
0            NaN      Diana    35.389999 2018-08-01      F   26   
1            NaN      Diana    21.200001 2018-08-01      F   26   
2            NaN    Michael     5.710000 2018-08-01      M   38   
3            NaN     Rhonda  2117.219971 2018-08-01      F   40   
4            NaN      Diana    17.950001 2018-08-01      F   26   

  merchant_suburb merchant_state          extraction     amount  \
0         Ashmore            QLD 2018-08-01 01:01:15  16.250000   
1          Sydney            NSW 2018-08-01 01:13:45  14.190000   
2          Sydney            NSW 2018-08-01 01:26:15   6.420000   
3         Buderim            QLD 2018-08-01 01:38:45  40.900002   
4   Mermaid Beach            QLD 2018-08-01 01:51:15   3.250000   

                     transaction_id    country     customer_id  \
0  a623070bfead4541a6b0fff8a09e706c  Australia  CUS-2487424745   
1  13270a2a902145da9db4c951e04b51b9  Australia  CUS-2487424745   
2  feb79e7ecd7048a5a36ec889d1a94270  Australia  CUS-2142601169   
3  2698170da3704fd981b15e64a006079e  Australia  CUS-1614226872   
4  329adf79878c4cf0aeb4188b4691c266  Australia  CUS-2487424745   

  merchant_long_lat movement  
0     153.38 -27.99    debit  
1     151.21 -33.87    debit  
2     151.21 -33.87    debit  
3     153.05 -26.68    debit  
4     153.44 -28.06    debit

In [124]:
df.extraction.dt.isocalendar().day

0        3
1        3
2        3
3        3
4        3
        ..
12038    3
12039    3
12040    3
12041    3
12042    3
Name: day, Length: 12043, dtype: UInt32

Timestamp('1999-01-01 00:00:00')

In [49]:
# Now the date needs to be dropped from the extraction column
import pandas
from io import StringIO

# load data 
df['extraction'] = pandas.to_datetime(df['extraction'])

# make the required change
without_date = df['extraction'].apply( lambda d : d.time() )
df['extraction'] = without_date

# display results
print(df)

           status  card_present_flag bpay_biller_code         account  \
0      authorized                1.0              NaN  ACC-1598451071   
1      authorized                0.0              NaN  ACC-1598451071   
2      authorized                1.0              NaN  ACC-1222300524   
3      authorized                1.0              NaN  ACC-1037050564   
4      authorized                1.0              NaN  ACC-1598451071   
...           ...                ...              ...             ...   
12038  authorized                0.0              NaN  ACC-3021093232   
12039  authorized                1.0              NaN  ACC-1608363396   
12040  authorized                1.0              NaN  ACC-3827517394   
12041  authorized                1.0              NaN  ACC-2920611728   
12042  authorized                1.0              NaN  ACC-1443681913   

      currency       long_lat txn_description  \
0          AUD  153.41 -27.95             POS   
1          AUD  153.41 -2

In [50]:
df.head()

status  card_present_flag bpay_biller_code         account currency  \
0  authorized                1.0              NaN  ACC-1598451071      AUD   
1  authorized                0.0              NaN  ACC-1598451071      AUD   
2  authorized                1.0              NaN  ACC-1222300524      AUD   
3  authorized                1.0              NaN  ACC-1037050564      AUD   
4  authorized                1.0              NaN  ACC-1598451071      AUD   

        long_lat txn_description                           merchant_id  \
0  153.41 -27.95             POS  81c48296-73be-44a7-befa-d053f48ce7cd   
1  153.41 -27.95       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e   
2  151.23 -33.94             POS  835c231d-8cdf-4e96-859d-e9d571760cf0   
3  153.10 -27.66       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673   
4  153.41 -27.95       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0   

   merchant_code first_name      balance       date gender  age  \
0            NaN      Diana    35.389999 2018-08-01      F   26   
1            NaN      Diana    21.200001 2018-08-01      F   26   
2            NaN    Michael     5.710000 2018-08-01      M   38   
3            NaN     Rhonda  2117.219971 2018-08-01      F   40   
4            NaN      Diana    17.950001 2018-08-01      F   26   

  merchant_suburb merchant_state extraction     amount  \
0         Ashmore            QLD   01:01:15  16.250000   
1          Sydney            NSW   01:13:45  14.190000   
2          Sydney            NSW   01:26:15   6.420000   
3         Buderim            QLD   01:38:45  40.900002   
4   Mermaid Beach            QLD   01:51:15   3.250000   

                     transaction_id    country     customer_id  \
0  a623070bfead4541a6b0fff8a09e706c  Australia  CUS-2487424745   
1  13270a2a902145da9db4c951e04b51b9  Australia  CUS-2487424745   
2  feb79e7ecd7048a5a36ec889d1a94270  Australia  CUS-2142601169   
3  2698170da3704fd981b15e64a006079e  Australia  CUS-1614226872   
4  329adf79878c4cf0aeb4188b4691c266  Australia  CUS-2487424745   

  merchant_long_lat movement  
0     153.38 -27.99    debit  
1     151.21 -33.87    debit  
2     151.21 -33.87    debit  
3     153.05 -26.68    debit  
4     153.44 -28.06    debit

In [51]:
# We now have extraction date and extraction time column. renaming for consistency
df.rename(columns = {'date':'extraction_date','extraction':'extraction_time'},inplace = True)

In [52]:
df.head()

status  card_present_flag bpay_biller_code         account currency  \
0  authorized                1.0              NaN  ACC-1598451071      AUD   
1  authorized                0.0              NaN  ACC-1598451071      AUD   
2  authorized                1.0              NaN  ACC-1222300524      AUD   
3  authorized                1.0              NaN  ACC-1037050564      AUD   
4  authorized                1.0              NaN  ACC-1598451071      AUD   

        long_lat txn_description                           merchant_id  \
0  153.41 -27.95             POS  81c48296-73be-44a7-befa-d053f48ce7cd   
1  153.41 -27.95       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e   
2  151.23 -33.94             POS  835c231d-8cdf-4e96-859d-e9d571760cf0   
3  153.10 -27.66       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673   
4  153.41 -27.95       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0   

   merchant_code first_name      balance extraction_date gender  age  \
0            NaN      Diana    35.389999      2018-08-01      F   26   
1            NaN      Diana    21.200001      2018-08-01      F   26   
2            NaN    Michael     5.710000      2018-08-01      M   38   
3            NaN     Rhonda  2117.219971      2018-08-01      F   40   
4            NaN      Diana    17.950001      2018-08-01      F   26   

  merchant_suburb merchant_state extraction_time     amount  \
0         Ashmore            QLD        01:01:15  16.250000   
1          Sydney            NSW        01:13:45  14.190000   
2          Sydney            NSW        01:26:15   6.420000   
3         Buderim            QLD        01:38:45  40.900002   
4   Mermaid Beach            QLD        01:51:15   3.250000   

                     transaction_id    country     customer_id  \
0  a623070bfead4541a6b0fff8a09e706c  Australia  CUS-2487424745   
1  13270a2a902145da9db4c951e04b51b9  Australia  CUS-2487424745   
2  feb79e7ecd7048a5a36ec889d1a94270  Australia  CUS-2142601169   
3  2698170da3704fd981b15e64a006079e  Australia  CUS-1614226872   
4  329adf79878c4cf0aeb4188b4691c266  Australia  CUS-2487424745   

  merchant_long_lat movement  
0     153.38 -27.99    debit  
1     151.21 -33.87    debit  
2     151.21 -33.87    debit  
3     153.05 -26.68    debit  
4     153.44 -28.06    debit

In [66]:
# rearranging the columns
headers = list(df.columns)
headers

['status',
 'card_present_flag',
 'bpay_biller_code',
 'account',
 'currency',
 'long_lat',
 'txn_description',
 'merchant_id',
 'merchant_code',
 'first_name',
 'balance',
 'merchant_state',
 'gender',
 'age',
 'merchant_suburb',
 'extraction_date',
 'extraction_time',
 'amount',
 'transaction_id',
 'country',
 'customer_id',
 'merchant_long_lat',
 'movement']

In [67]:
for index, column in enumerate(headers):
    print(index,column)

0 status
1 card_present_flag
2 bpay_biller_code
3 account
4 currency
5 long_lat
6 txn_description
7 merchant_id
8 merchant_code
9 first_name
10 balance
11 merchant_state
12 gender
13 age
14 merchant_suburb
15 extraction_date
16 extraction_time
17 amount
18 transaction_id
19 country
20 customer_id
21 merchant_long_lat
22 movement


In [68]:
df = df[headers]

In [70]:
# the date and time and now located next to each other
df

status  card_present_flag bpay_biller_code         account  \
0      authorized                1.0              NaN  ACC-1598451071   
1      authorized                0.0              NaN  ACC-1598451071   
2      authorized                1.0              NaN  ACC-1222300524   
3      authorized                1.0              NaN  ACC-1037050564   
4      authorized                1.0              NaN  ACC-1598451071   
...           ...                ...              ...             ...   
12038  authorized                0.0              NaN  ACC-3021093232   
12039  authorized                1.0              NaN  ACC-1608363396   
12040  authorized                1.0              NaN  ACC-3827517394   
12041  authorized                1.0              NaN  ACC-2920611728   
12042  authorized                1.0              NaN  ACC-1443681913   

      currency       long_lat txn_description  \
0          AUD  153.41 -27.95             POS   
1          AUD  153.41 -27.95       SALES-POS   
2          AUD  151.23 -33.94             POS   
3          AUD  153.10 -27.66       SALES-POS   
4          AUD  153.41 -27.95       SALES-POS   
...        ...            ...             ...   
12038      AUD  149.83 -29.47             POS   
12039      AUD  151.22 -33.87       SALES-POS   
12040      AUD  151.12 -33.89             POS   
12041      AUD  144.96 -37.76       SALES-POS   
12042      AUD  150.92 -33.77       SALES-POS   

                                merchant_id  merchant_code first_name  \
0      81c48296-73be-44a7-befa-d053f48ce7cd            NaN      Diana   
1      830a451c-316e-4a6a-bf25-e37caedca49e            NaN      Diana   
2      835c231d-8cdf-4e96-859d-e9d571760cf0            NaN    Michael   
3      48514682-c78a-4a88-b0da-2d6302e64673            NaN     Rhonda   
4      b4e02c10-0852-4273-b8fd-7b3395e32eb0            NaN      Diana   
...                                     ...            ...        ...   
12038  32aa73dc-b7c2-4161-b14d-6271b96ce792            NaN    Melissa   
12039  296a0500-8552-48ac-ac81-ec37065b568e            NaN     Robert   
12040  e5975ab4-08f7-4725-a369-24cc0e35ed6e            NaN      Craig   
12041  af49051a-591d-4b08-bd3c-27730b70ed37            NaN      Tyler   
12042  f31f4b14-2040-40ec-a120-b141bb274cbd            NaN       Ryan   

            balance merchant_state gender  age merchant_suburb  \
0         35.389999            QLD      F   26         Ashmore   
1         21.200001            NSW      F   26          Sydney   
2          5.710000            NSW      M   38          Sydney   
3       2117.219971            QLD      F   40         Buderim   
4         17.950001            QLD      F   26   Mermaid Beach   
...             ...            ...    ...  ...             ...   
12038  14054.139648            VIC      F   30        Ringwood   
12039   9137.790039            NSW      M   20          Casula   
12040  45394.570312            NSW      M   28      Kings Park   
12041  11350.669922            VIC      M   69        Oakleigh   
12042   5517.910156            NSW      M   31          Mascot   

      extraction_date extraction_time     amount  \
0          2018-08-01        01:01:15  16.250000   
1          2018-08-01        01:13:45  14.190000   
2          2018-08-01        01:26:15   6.420000   
3          2018-08-01        01:38:45  40.900002   
4          2018-08-01        01:51:15   3.250000   
...               ...             ...        ...   
12038      2018-10-31        23:09:06   9.790000   
12039      2018-10-31        23:21:46  63.869999   
12040      2018-10-31        23:34:25  43.959999   
12041      2018-10-31        23:47:05  30.770000   
12042      2018-10-31        23:59:44  22.360001   

                         transaction_id    country     customer_id  \
0      a623070bfead4541a6b0fff8a09e706c  Australia  CUS-2487424745   
1      13270a2a902145da9db4c951e04b51b9  Australia  CUS-2487424745   
2      feb79e7ecd7048a5a36ec889d1a94270 

Looking at the dataframe as is, there are at least 2 columns that appear to be NaN values, 'bpay_biller_code' and 'merchant_code'.

In [71]:
# Expanding all columns to check for other possible NaN columns
pd.set_option('max_columns', None)
pd.set_option('min_rows',25)
df

status  card_present_flag bpay_biller_code         account  \
0      authorized                1.0              NaN  ACC-1598451071   
1      authorized                0.0              NaN  ACC-1598451071   
2      authorized                1.0              NaN  ACC-1222300524   
3      authorized                1.0              NaN  ACC-1037050564   
4      authorized                1.0              NaN  ACC-1598451071   
5          posted                NaN              NaN  ACC-1608363396   
6      authorized                1.0              NaN  ACC-2776252858   
7      authorized                1.0              NaN  ACC-2776252858   
8      authorized                1.0              NaN   ACC-182446574   
9          posted                NaN              NaN   ACC-602667573   
10         posted                NaN              NaN  ACC-2171593283   
11         posted                NaN              NaN  ACC-2776252858   
...           ...                ...              ...             ...   
12031  authorized                0.0              NaN  ACC-1443681913   
12032      posted                NaN              NaN  ACC-1710017148   
12033  authorized                1.0              NaN  ACC-3100725361   
12034  authorized                1.0              NaN  ACC-1516130869   
12035      posted                NaN              NaN  ACC-1973887809   
12036      posted                NaN              NaN  ACC-1349834573   
12037      posted                NaN              NaN   ACC-240804743   
12038  authorized                0.0              NaN  ACC-3021093232   
12039  authorized                1.0              NaN  ACC-1608363396   
12040  authorized                1.0              NaN  ACC-3827517394   
12041  authorized                1.0              NaN  ACC-2920611728   
12042  authorized                1.0              NaN  ACC-1443681913   

      currency       long_lat txn_description  \
0          AUD  153.41 -27.95             POS   
1          AUD  153.41 -27.95       SALES-POS   
2          AUD  151.23 -33.94             POS   
3          AUD  153.10 -27.66       SALES-POS   
4          AUD  153.41 -27.95       SALES-POS   
5          AUD  151.22 -33.87         PAYMENT   
6          AUD  144.95 -37.76       SALES-POS   
7          AUD  144.95 -37.76             POS   
8          AUD  116.06 -32.00             POS   
9          AUD  151.23 -33.96      INTER BANK   
10         AUD  146.94 -36.04         PAYMENT   
11         AUD  144.95 -37.76         PAYMENT   
...        ...            ...             ...   
12031      AUD  150.92 -33.77       SALES-POS   
12032      AUD  150.82 -34.01         PAYMENT   
12033      AUD  145.73 -17.03       SALES-POS   
12034      AUD  130.84 -12.45             POS   
12035      AUD  115.78 -31.90         PAYMENT   
12036      AUD  145.17 -37.84         PAYMENT   
12037      AUD  121.48 -30.75         PAYMENT   
12038      AUD  149.83 -29.47             POS   
12039      AUD  151.22 -33.87       SALES-POS   
12040      AUD  151.12 -33.89             POS   
12041      AUD  144.96 -37.76       SALES-POS   
12042      AUD  150.92 -33.77       SALES-POS   

                                merchant_id  merchant_code first_name  \
0      81c48296-73be-44a7-befa-d053f48ce7cd            NaN      Diana   
1      830a451c-316e-4a6a-bf25-e37caedca49e            NaN      Diana   
2      835c231d-8cdf-4e96-859d-e9d571760cf0            NaN    Michael   
3      48514682-c78a-4a88-b0da-2d6302e64673            NaN     Rhonda   
4      b4e02c10-0852-4273-b8fd-7b3395e32eb0            NaN      Diana   
5                                       NaN            NaN     Robert   
6      3aa18033-a0a9-4190-a117-b7caaf4d07e3            NaN    Kristin   
7      ee58145d-26e8-4b01-9cd9-62378014f21d            NaN    Kristin   
8      cfbf535e-caa8-499f-9d41-bbdc2b836d85            NaN      Tonya   
9                                       NaN            NaN    Michael   
10                                

There appear no other immediate only NaN value colums

In [72]:
# some information on the dtype, column and row totals
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12043 entries, 0 to 12042
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   status             12043 non-null  object        
 1   card_present_flag  7717 non-null   float32       
 2   bpay_biller_code   885 non-null    object        
 3   account            12043 non-null  object        
 4   currency           12043 non-null  object        
 5   long_lat           12043 non-null  object        
 6   txn_description    12043 non-null  object        
 7   merchant_id        7717 non-null   object        
 8   merchant_code      883 non-null    float32       
 9   first_name         12043 non-null  object        
 10  balance            12043 non-null  float32       
 11  merchant_state     7717 non-null   object        
 12  gender             12043 non-null  object        
 13  age                12043 non-null  int8          
 14  mercha

In [73]:
# checking for total null values in all of the rows
df.isnull().sum()

status                   0
card_present_flag     4326
bpay_biller_code     11158
account                  0
currency                 0
long_lat                 0
txn_description          0
merchant_id           4326
merchant_code        11160
first_name               0
balance                  0
merchant_state        4326
gender                   0
age                      0
merchant_suburb       4326
extraction_date          0
extraction_time          0
amount                   0
transaction_id           0
country                  0
customer_id              0
merchant_long_lat     4326
movement                 0
dtype: int64

From this information we can notice that two rows are almost entirely null values. out of 12042 rows, 'bpay_biller_code' has 11158 null values and 'merchant_code' has 11160 null values. Both of these rows have nearly 90% missing data. From this we can conclude that these two rows can be dropped.

4 other rows have exactly the same amount of null values and we can explore this a bit further.

In [74]:
# dropping both rows and creating a new dataframe
df1 = df.drop(['bpay_biller_code', 'merchant_code'], axis = 1)
df1

status  card_present_flag         account currency       long_lat  \
0      authorized                1.0  ACC-1598451071      AUD  153.41 -27.95   
1      authorized                0.0  ACC-1598451071      AUD  153.41 -27.95   
2      authorized                1.0  ACC-1222300524      AUD  151.23 -33.94   
3      authorized                1.0  ACC-1037050564      AUD  153.10 -27.66   
4      authorized                1.0  ACC-1598451071      AUD  153.41 -27.95   
5          posted                NaN  ACC-1608363396      AUD  151.22 -33.87   
6      authorized                1.0  ACC-2776252858      AUD  144.95 -37.76   
7      authorized                1.0  ACC-2776252858      AUD  144.95 -37.76   
8      authorized                1.0   ACC-182446574      AUD  116.06 -32.00   
9          posted                NaN   ACC-602667573      AUD  151.23 -33.96   
10         posted                NaN  ACC-2171593283      AUD  146.94 -36.04   
11         posted                NaN  ACC-2776252858      AUD  144.95 -37.76   
...           ...                ...             ...      ...            ...   
12031  authorized                0.0  ACC-1443681913      AUD  150.92 -33.77   
12032      posted                NaN  ACC-1710017148      AUD  150.82 -34.01   
12033  authorized                1.0  ACC-3100725361      AUD  145.73 -17.03   
12034  authorized                1.0  ACC-1516130869      AUD  130.84 -12.45   
12035      posted                NaN  ACC-1973887809      AUD  115.78 -31.90   
12036      posted                NaN  ACC-1349834573      AUD  145.17 -37.84   
12037      posted                NaN   ACC-240804743      AUD  121.48 -30.75   
12038  authorized                0.0  ACC-3021093232      AUD  149.83 -29.47   
12039  authorized                1.0  ACC-1608363396      AUD  151.22 -33.87   
12040  authorized                1.0  ACC-3827517394      AUD  151.12 -33.89   
12041  authorized                1.0  ACC-2920611728      AUD  144.96 -37.76   
12042  authorized                1.0  ACC-1443681913      AUD  150.92 -33.77   

      txn_description                           merchant_id first_name  \
0                 POS  81c48296-73be-44a7-befa-d053f48ce7cd      Diana   
1           SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e      Diana   
2                 POS  835c231d-8cdf-4e96-859d-e9d571760cf0    Michael   
3           SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673     Rhonda   
4           SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0      Diana   
5             PAYMENT                                   NaN     Robert   
6           SALES-POS  3aa18033-a0a9-4190-a117-b7caaf4d07e3    Kristin   
7                 POS  ee58145d-26e8-4b01-9cd9-62378014f21d    Kristin   
8                 POS  cfbf535e-caa8-499f-9d41-bbdc2b836d85      Tonya   
9          INTER BANK                                   NaN    Michael   
10            PAYMENT                                   NaN   Fernando   
11            PAYMENT                                   NaN    Kristin   
...               ...                                   ...        ...   
12031       SALES-POS  6fcdc956-3548-40b2-a2ce-9dbc6cb64eb9       Ryan   
12032         PAYMENT                                   NaN   Michelle   
12033       SALES-POS  f17a5cbd-fa5d-47db-8235-c7f449597dd0     Ronald   
12034             POS  7a3610ce-2f3c-4c7f-becf-987ddeb1c533    Jessica   
12035         PAYMENT                                   NaN    Antonio   
12036         PAYMENT                                   NaN    Matthew   
12037         PAYMENT                                   NaN    Kenneth   
12038             POS  32aa73dc-b7c2-4161-b14d-6271b96ce792    Melissa   
12039       SALES-POS  296a0500-8552-48ac-ac81-ec37065b568e     Robert   
12040             POS  e5975ab4-08f7-4725-a369-24cc0e35ed6e      Craig   
12041       SALES-POS  af49051a-591d-4b08-bd3c-27730b70ed37      Tyler   
12042       SALES-POS  f31f4b14-2040-40ec-a120-b141bb274cbd       Ryan   

     

In [75]:
# checking to see if any currency besides AUD was used
df1[df1['currency'] != 'AUD']

Empty DataFrame
Columns: [status, card_present_flag, account, currency, long_lat, txn_description, merchant_id, first_name, balance, merchant_state, gender, age, merchant_suburb, extraction_date, extraction_time, amount, transaction_id, country, customer_id, merchant_long_lat, movement]
Index: []

In [76]:
# checking if any purchases were made outside of Australia
df1[df1['country'] != 'Australia']

Empty DataFrame
Columns: [status, card_present_flag, account, currency, long_lat, txn_description, merchant_id, first_name, balance, merchant_state, gender, age, merchant_suburb, extraction_date, extraction_time, amount, transaction_id, country, customer_id, merchant_long_lat, movement]
Index: []

Both of in these searches yielded 0 results which means that all the values in both of the checked columuns are the same. We know that all currency is in AUD and all transactions were made within Australia. These two columns provide no further insight and they can be dropped

In [77]:
# Dropping currency and country columns
df1.drop(['country','currency'], axis = 1, inplace = True)

In [78]:
# Check the resulting dataframe
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12043 entries, 0 to 12042
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   status             12043 non-null  object        
 1   card_present_flag  7717 non-null   float32       
 2   account            12043 non-null  object        
 3   long_lat           12043 non-null  object        
 4   txn_description    12043 non-null  object        
 5   merchant_id        7717 non-null   object        
 6   first_name         12043 non-null  object        
 7   balance            12043 non-null  float32       
 8   merchant_state     7717 non-null   object        
 9   gender             12043 non-null  object        
 10  age                12043 non-null  int8          
 11  merchant_suburb    7717 non-null   object        
 12  extraction_date    12043 non-null  datetime64[ns]
 13  extraction_time    12043 non-null  object        
 14  amount

In [79]:
df1.head()

status  card_present_flag         account       long_lat  \
0  authorized                1.0  ACC-1598451071  153.41 -27.95   
1  authorized                0.0  ACC-1598451071  153.41 -27.95   
2  authorized                1.0  ACC-1222300524  151.23 -33.94   
3  authorized                1.0  ACC-1037050564  153.10 -27.66   
4  authorized                1.0  ACC-1598451071  153.41 -27.95   

  txn_description                           merchant_id first_name  \
0             POS  81c48296-73be-44a7-befa-d053f48ce7cd      Diana   
1       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e      Diana   
2             POS  835c231d-8cdf-4e96-859d-e9d571760cf0    Michael   
3       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673     Rhonda   
4       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0      Diana   

       balance merchant_state gender  age merchant_suburb extraction_date  \
0    35.389999            QLD      F   26         Ashmore      2018-08-01   
1    21.200001            NSW      F   26          Sydney      2018-08-01   
2     5.710000            NSW      M   38          Sydney      2018-08-01   
3  2117.219971            QLD      F   40         Buderim      2018-08-01   
4    17.950001            QLD      F   26   Mermaid Beach      2018-08-01   

  extraction_time     amount                    transaction_id  \
0        01:01:15  16.250000  a623070bfead4541a6b0fff8a09e706c   
1        01:13:45  14.190000  13270a2a902145da9db4c951e04b51b9   
2        01:26:15   6.420000  feb79e7ecd7048a5a36ec889d1a94270   
3        01:38:45  40.900002  2698170da3704fd981b15e64a006079e   
4        01:51:15   3.250000  329adf79878c4cf0aeb4188b4691c266   

      customer_id merchant_long_lat movement  
0  CUS-2487424745     153.38 -27.99    debit  
1  CUS-2487424745     151.21 -33.87    debit  
2  CUS-2142601169     151.21 -33.87    debit  
3  CUS-1614226872     153.05 -26.68    debit  
4  CUS-2487424745     153.44 -28.06    debit

In [80]:
# The transaction_id column seems to be a randomly generated value consisting of letters and numbers. 
# I will check for duplicates to make sure this is the case

dupes = df1.duplicated(subset = 'transaction_id').sum()
dupes

0

There are no matching values in the transaction_id column. This  indicates that the values are all different for each transaction regardles of whether the same customer is making another purchase. So I will drop this column as it will not give us any information.

In [81]:
# dropping the transaction_id column
df1.drop('transaction_id', axis = 1, inplace = True)
df1.head()

status  card_present_flag         account       long_lat  \
0  authorized                1.0  ACC-1598451071  153.41 -27.95   
1  authorized                0.0  ACC-1598451071  153.41 -27.95   
2  authorized                1.0  ACC-1222300524  151.23 -33.94   
3  authorized                1.0  ACC-1037050564  153.10 -27.66   
4  authorized                1.0  ACC-1598451071  153.41 -27.95   

  txn_description                           merchant_id first_name  \
0             POS  81c48296-73be-44a7-befa-d053f48ce7cd      Diana   
1       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e      Diana   
2             POS  835c231d-8cdf-4e96-859d-e9d571760cf0    Michael   
3       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673     Rhonda   
4       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0      Diana   

       balance merchant_state gender  age merchant_suburb extraction_date  \
0    35.389999            QLD      F   26         Ashmore      2018-08-01   
1    21.200001            NSW      F   26          Sydney      2018-08-01   
2     5.710000            NSW      M   38          Sydney      2018-08-01   
3  2117.219971            QLD      F   40         Buderim      2018-08-01   
4    17.950001            QLD      F   26   Mermaid Beach      2018-08-01   

  extraction_time     amount     customer_id merchant_long_lat movement  
0        01:01:15  16.250000  CUS-2487424745     153.38 -27.99    debit  
1        01:13:45  14.190000  CUS-2487424745     151.21 -33.87    debit  
2        01:26:15   6.420000  CUS-2142601169     151.21 -33.87    debit  
3        01:38:45  40.900002  CUS-1614226872     153.05 -26.68    debit  
4        01:51:15   3.250000  CUS-2487424745     153.44 -28.06    debit

In [82]:
# Privacy is important. It appears that the customer_id, first_name and account are all linked 
# Therefore I will drop the name and the account column as we can use the customer_id for analysis 

# first I will group by first_name to check
df1.groupby(['first_name'])['account', 'customer_id'].count()

<ipython-input-82-8349221b96a4>:5: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df1.groupby(['first_name'])['account', 'customer_id'].count()


account  customer_id
first_name                      
Abigail         102          102
Alexander        91           91
Amy             124          124
Andrew           34           34
Antonio         115          115
Ashley           75           75
Barry           122          122
Billy            77           77
Brian            63           63
Catherine        62           62
Charles         145          145
Christine       139          139
...             ...          ...
Ryan            239          239
Sandra          174          174
Sarah            86           86
Scott           109          109
Stephanie        73           73
Susan           282          282
Tiffany         230          230
Tim              78           78
Timothy          72           72
Tonya           292          292
Tyler           218          218
Virginia        199          199

[80 rows x 2 columns]

As hypothesized, each customer has an account and customer_id linked to their name. 

In [83]:
# dropping the account and first_name columns for privacy purposes
df1.drop(['account', 'first_name'], axis = 1, inplace = True)

In [84]:
df1.head()

status  card_present_flag       long_lat txn_description  \
0  authorized                1.0  153.41 -27.95             POS   
1  authorized                0.0  153.41 -27.95       SALES-POS   
2  authorized                1.0  151.23 -33.94             POS   
3  authorized                1.0  153.10 -27.66       SALES-POS   
4  authorized                1.0  153.41 -27.95       SALES-POS   

                            merchant_id      balance merchant_state gender  \
0  81c48296-73be-44a7-befa-d053f48ce7cd    35.389999            QLD      F   
1  830a451c-316e-4a6a-bf25-e37caedca49e    21.200001            NSW      F   
2  835c231d-8cdf-4e96-859d-e9d571760cf0     5.710000            NSW      M   
3  48514682-c78a-4a88-b0da-2d6302e64673  2117.219971            QLD      F   
4  b4e02c10-0852-4273-b8fd-7b3395e32eb0    17.950001            QLD      F   

   age merchant_suburb extraction_date extraction_time     amount  \
0   26         Ashmore      2018-08-01        01:01:15  16.250000   
1   26          Sydney      2018-08-01        01:13:45  14.190000   
2   38          Sydney      2018-08-01        01:26:15   6.420000   
3   40         Buderim      2018-08-01        01:38:45  40.900002   
4   26   Mermaid Beach      2018-08-01        01:51:15   3.250000   

      customer_id merchant_long_lat movement  
0  CUS-2487424745     153.38 -27.99    debit  
1  CUS-2487424745     151.21 -33.87    debit  
2  CUS-2142601169     151.21 -33.87    debit  
3  CUS-1614226872     153.05 -26.68    debit  
4  CUS-2487424745     153.44 -28.06    debit

In [85]:
# separating the long_lat column into two new columns
df['long'] = df['long_lat'].apply(lambda x: x.split('-')[0]).astype('float32')
df['lat'] = df['long_lat'].apply(lambda x: x.split('-')[1]).astype('float32')


### Back to creating a data frame of the 4 rows with the same null values


In [86]:
equalnull_df = df1[['card_present_flag','merchant_id','merchant_suburb', 'merchant_state','merchant_long_lat' ]]

In [87]:
equalnull_df

card_present_flag                           merchant_id  \
0                    1.0  81c48296-73be-44a7-befa-d053f48ce7cd   
1                    0.0  830a451c-316e-4a6a-bf25-e37caedca49e   
2                    1.0  835c231d-8cdf-4e96-859d-e9d571760cf0   
3                    1.0  48514682-c78a-4a88-b0da-2d6302e64673   
4                    1.0  b4e02c10-0852-4273-b8fd-7b3395e32eb0   
5                    NaN                                   NaN   
6                    1.0  3aa18033-a0a9-4190-a117-b7caaf4d07e3   
7                    1.0  ee58145d-26e8-4b01-9cd9-62378014f21d   
8                    1.0  cfbf535e-caa8-499f-9d41-bbdc2b836d85   
9                    NaN                                   NaN   
10                   NaN                                   NaN   
11                   NaN                                   NaN   
...                  ...                                   ...   
12031                0.0  6fcdc956-3548-40b2-a2ce-9dbc6cb64eb9   
12032                NaN                                   NaN   
12033                1.0  f17a5cbd-fa5d-47db-8235-c7f449597dd0   
12034                1.0  7a3610ce-2f3c-4c7f-becf-987ddeb1c533   
12035                NaN                                   NaN   
12036                NaN                                   NaN   
12037                NaN                                   NaN   
12038                0.0  32aa73dc-b7c2-4161-b14d-6271b96ce792   
12039                1.0  296a0500-8552-48ac-ac81-ec37065b568e   
12040                1.0  e5975ab4-08f7-4725-a369-24cc0e35ed6e   
12041                1.0  af49051a-591d-4b08-bd3c-27730b70ed37   
12042                1.0  f31f4b14-2040-40ec-a120-b141bb274cbd   

      merchant_suburb merchant_state merchant_long_lat  
0             Ashmore            QLD     153.38 -27.99  
1              Sydney            NSW     151.21 -33.87  
2              Sydney            NSW     151.21 -33.87  
3             Buderim            QLD     153.05 -26.68  
4       Mermaid Beach            QLD     153.44 -28.06  
5                 NaN            NaN               NaN  
6            Kalkallo            VIC     144.95 -37.53  
7           Melbourne            VIC     144.96 -37.81  
8              Yokine             WA      115.85 -31.9  
9                 NaN            NaN               NaN  
10                NaN            NaN               NaN  
11                NaN            NaN               NaN  
...               ...            ...               ...  
12031    Marrickville            NSW     151.17 -33.91  
12032             NaN            NaN               NaN  
12033      Smithfield            QLD     145.69 -16.84  
12034     Darwin City             NT     130.84 -12.46  
12035             NaN            NaN               NaN  
12036             NaN            NaN               NaN  
12037             NaN            NaN               NaN  
12038        Ringwood            VIC     145.23 -37.81  
12039          Casula            NSW     150.88 -33.96  
12040      Kings Park            NSW     150.92 -33.74  
12041        Oakleigh            VIC     145.09 -37.91  
12042          Mascot            NSW     151.19 -33.93  

[12043 rows x 5 columns]

In [88]:
# changing the number of rows that appear to look for insight as to why all 4 columns have the same amount of null values
pd.set_option('display.min_rows', 25)
equalnull_df

card_present_flag                           merchant_id  \
0                    1.0  81c48296-73be-44a7-befa-d053f48ce7cd   
1                    0.0  830a451c-316e-4a6a-bf25-e37caedca49e   
2                    1.0  835c231d-8cdf-4e96-859d-e9d571760cf0   
3                    1.0  48514682-c78a-4a88-b0da-2d6302e64673   
4                    1.0  b4e02c10-0852-4273-b8fd-7b3395e32eb0   
5                    NaN                                   NaN   
6                    1.0  3aa18033-a0a9-4190-a117-b7caaf4d07e3   
7                    1.0  ee58145d-26e8-4b01-9cd9-62378014f21d   
8                    1.0  cfbf535e-caa8-499f-9d41-bbdc2b836d85   
9                    NaN                                   NaN   
10                   NaN                                   NaN   
11                   NaN                                   NaN   
...                  ...                                   ...   
12031                0.0  6fcdc956-3548-40b2-a2ce-9dbc6cb64eb9   
12032                NaN                                   NaN   
12033                1.0  f17a5cbd-fa5d-47db-8235-c7f449597dd0   
12034                1.0  7a3610ce-2f3c-4c7f-becf-987ddeb1c533   
12035                NaN                                   NaN   
12036                NaN                                   NaN   
12037                NaN                                   NaN   
12038                0.0  32aa73dc-b7c2-4161-b14d-6271b96ce792   
12039                1.0  296a0500-8552-48ac-ac81-ec37065b568e   
12040                1.0  e5975ab4-08f7-4725-a369-24cc0e35ed6e   
12041                1.0  af49051a-591d-4b08-bd3c-27730b70ed37   
12042                1.0  f31f4b14-2040-40ec-a120-b141bb274cbd   

      merchant_suburb merchant_state merchant_long_lat  
0             Ashmore            QLD     153.38 -27.99  
1              Sydney            NSW     151.21 -33.87  
2              Sydney            NSW     151.21 -33.87  
3             Buderim            QLD     153.05 -26.68  
4       Mermaid Beach            QLD     153.44 -28.06  
5                 NaN            NaN               NaN  
6            Kalkallo            VIC     144.95 -37.53  
7           Melbourne            VIC     144.96 -37.81  
8              Yokine             WA      115.85 -31.9  
9                 NaN            NaN               NaN  
10                NaN            NaN               NaN  
11                NaN            NaN               NaN  
...               ...            ...               ...  
12031    Marrickville            NSW     151.17 -33.91  
12032             NaN            NaN               NaN  
12033      Smithfield            QLD     145.69 -16.84  
12034     Darwin City             NT     130.84 -12.46  
12035             NaN            NaN               NaN  
12036             NaN            NaN               NaN  
12037             NaN            NaN               NaN  
12038        Ringwood            VIC     145.23 -37.81  
12039          Casula            NSW     150.88 -33.96  
12040      Kings Park            NSW     150.92 -33.74  
12041        Oakleigh            VIC     145.09 -37.91  
12042          Mascot            NSW     151.19 -33.93  

[12043 rows x 5 columns]

Looking at this data, we can tell that all four columns have null values in the same rows. 

--------

### Returning to the original dataframe 

In [89]:
df1

status  card_present_flag       long_lat txn_description  \
0      authorized                1.0  153.41 -27.95             POS   
1      authorized                0.0  153.41 -27.95       SALES-POS   
2      authorized                1.0  151.23 -33.94             POS   
3      authorized                1.0  153.10 -27.66       SALES-POS   
4      authorized                1.0  153.41 -27.95       SALES-POS   
5          posted                NaN  151.22 -33.87         PAYMENT   
6      authorized                1.0  144.95 -37.76       SALES-POS   
7      authorized                1.0  144.95 -37.76             POS   
8      authorized                1.0  116.06 -32.00             POS   
9          posted                NaN  151.23 -33.96      INTER BANK   
10         posted                NaN  146.94 -36.04         PAYMENT   
11         posted                NaN  144.95 -37.76         PAYMENT   
...           ...                ...            ...             ...   
12031  authorized                0.0  150.92 -33.77       SALES-POS   
12032      posted                NaN  150.82 -34.01         PAYMENT   
12033  authorized                1.0  145.73 -17.03       SALES-POS   
12034  authorized                1.0  130.84 -12.45             POS   
12035      posted                NaN  115.78 -31.90         PAYMENT   
12036      posted                NaN  145.17 -37.84         PAYMENT   
12037      posted                NaN  121.48 -30.75         PAYMENT   
12038  authorized                0.0  149.83 -29.47             POS   
12039  authorized                1.0  151.22 -33.87       SALES-POS   
12040  authorized                1.0  151.12 -33.89             POS   
12041  authorized                1.0  144.96 -37.76       SALES-POS   
12042  authorized                1.0  150.92 -33.77       SALES-POS   

                                merchant_id       balance merchant_state  \
0      81c48296-73be-44a7-befa-d053f48ce7cd     35.389999            QLD   
1      830a451c-316e-4a6a-bf25-e37caedca49e     21.200001            NSW   
2      835c231d-8cdf-4e96-859d-e9d571760cf0      5.710000            NSW   
3      48514682-c78a-4a88-b0da-2d6302e64673   2117.219971            QLD   
4      b4e02c10-0852-4273-b8fd-7b3395e32eb0     17.950001            QLD   
5                                       NaN   1705.430054            NaN   
6      3aa18033-a0a9-4190-a117-b7caaf4d07e3   1248.359985            VIC   
7      ee58145d-26e8-4b01-9cd9-62378014f21d   1232.750000            VIC   
8      cfbf535e-caa8-499f-9d41-bbdc2b836d85    213.160004             WA   
9                                       NaN    466.579987            NaN   
10                                      NaN   4348.500000            NaN   
11                                      NaN   1203.750000            NaN   
...                                     ...           ...            ...   
12031  6fcdc956-3548-40b2-a2ce-9dbc6cb64eb9   5540.270020            NSW   
12032                                   NaN  20699.900391            NaN   
12033  f17a5cbd-fa5d-47db-8235-c7f449597dd0  10445.389648            QLD   
12034  7a3610ce-2f3c-4c7f-becf-987ddeb1c533   4946.620117             NT   
12035                                   NaN   9349.509766            NaN   
12036                                   NaN  18127.500000            NaN   
12037                                   NaN  24775.660156            NaN   
12038  32aa73dc-b7c2-4161-b14d-6271b96ce792  14054.139648            VIC   
12039  296a0500-8552-48ac-ac81-ec37065b568e   9137.790039            NSW   
12040  e5975ab4-08f7-4725-a369-24cc0e35ed6e  45394.570312            NSW   
12041  af49051a-591d-4b08-bd3c-27730b70ed37  11350.669922            VIC   
12042  f31f4b14-2040-40ec-a120-b141bb274cbd   5517.910156            NSW   

      gender  age merchant_suburb extraction_date extraction_time      amount  \
0          F   26         Ashmore      2018-08-01        01:01:15   16.250000   
1          F   26        

# 

In [90]:
# checking to see if column status = posted has a card_present_value of NaN
pd.reset_option('display.min_rows')
df1[df1['status'] == 'posted']


status  card_present_flag       long_lat txn_description merchant_id  \
5      posted                NaN  151.22 -33.87         PAYMENT         NaN   
9      posted                NaN  151.23 -33.96      INTER BANK         NaN   
10     posted                NaN  146.94 -36.04         PAYMENT         NaN   
11     posted                NaN  144.95 -37.76         PAYMENT         NaN   
13     posted                NaN  151.27 -33.76      INTER BANK         NaN   
...       ...                ...            ...             ...         ...   
12027  posted                NaN  144.71 -37.58      PHONE BANK         NaN   
12032  posted                NaN  150.82 -34.01         PAYMENT         NaN   
12035  posted                NaN  115.78 -31.90         PAYMENT         NaN   
12036  posted                NaN  145.17 -37.84         PAYMENT         NaN   
12037  posted                NaN  121.48 -30.75         PAYMENT         NaN   

            balance merchant_state gender  age merchant_suburb  \
5       1705.430054            NaN      M   20             NaN   
9        466.579987            NaN      M   40             NaN   
10      4348.500000            NaN      M   19             NaN   
11      1203.750000            NaN      F   43             NaN   
13      4438.160156            NaN      M   23             NaN   
...             ...            ...    ...  ...             ...   
12027  13537.750000            NaN      M   27             NaN   
12032  20699.900391            NaN      F   19             NaN   
12035   9349.509766            NaN      M   26             NaN   
12036  18127.500000            NaN      M   39             NaN   
12037  24775.660156            NaN      M   29             NaN   

      extraction_date extraction_time  amount     customer_id  \
5          2018-08-01        02:00:00   163.0  CUS-2688605418   
9          2018-08-01        06:00:00    21.0  CUS-2031327464   
10         2018-08-01        06:00:00    27.0  CUS-2317998716   
11         2018-08-01        06:00:00    29.0  CUS-4123612273   
13         2018-08-01        07:00:00    25.0  CUS-1462656821   
...               ...             ...     ...             ...   
12027      2018-10-31        21:00:00    67.0  CUS-3180318393   
12032      2018-10-31        22:00:00    33.0   CUS-883482547   
12035      2018-10-31        23:00:00    29.0  CUS-1646621553   
12036      2018-10-31        23:00:00    32.0  CUS-1790886359   
12037      2018-10-31        23:00:00    39.0   CUS-261674136   

      merchant_long_lat movement  
5                   NaN    debit  
9                   NaN    debit  
10                  NaN    debit  
11                  NaN    debit  
13                  NaN    debit  
...                 ...      ...  
12027               NaN    debit  
12032               NaN    debit  
12035               NaN    debit  
12036               NaN    debit  
12037               NaN    debit  

[4326 rows x 16 columns]

From this it appears that all status = posted have card_present_flag of NaN. For now, we will not be dropping these rows because the values from each row in other columns can still be used for insight. I will create a dataframe where all NaN values are dropped. This will get rid of 4326 rows. out of 12402 rows, this is nearly 35% of the data. 

In [91]:
df2 = df1.dropna()
df2

status  card_present_flag       long_lat txn_description  \
0      authorized                1.0  153.41 -27.95             POS   
1      authorized                0.0  153.41 -27.95       SALES-POS   
2      authorized                1.0  151.23 -33.94             POS   
3      authorized                1.0  153.10 -27.66       SALES-POS   
4      authorized                1.0  153.41 -27.95       SALES-POS   
...           ...                ...            ...             ...   
12038  authorized                0.0  149.83 -29.47             POS   
12039  authorized                1.0  151.22 -33.87       SALES-POS   
12040  authorized                1.0  151.12 -33.89             POS   
12041  authorized                1.0  144.96 -37.76       SALES-POS   
12042  authorized                1.0  150.92 -33.77       SALES-POS   

                                merchant_id       balance merchant_state  \
0      81c48296-73be-44a7-befa-d053f48ce7cd     35.389999            QLD   
1      830a451c-316e-4a6a-bf25-e37caedca49e     21.200001            NSW   
2      835c231d-8cdf-4e96-859d-e9d571760cf0      5.710000            NSW   
3      48514682-c78a-4a88-b0da-2d6302e64673   2117.219971            QLD   
4      b4e02c10-0852-4273-b8fd-7b3395e32eb0     17.950001            QLD   
...                                     ...           ...            ...   
12038  32aa73dc-b7c2-4161-b14d-6271b96ce792  14054.139648            VIC   
12039  296a0500-8552-48ac-ac81-ec37065b568e   9137.790039            NSW   
12040  e5975ab4-08f7-4725-a369-24cc0e35ed6e  45394.570312            NSW   
12041  af49051a-591d-4b08-bd3c-27730b70ed37  11350.669922            VIC   
12042  f31f4b14-2040-40ec-a120-b141bb274cbd   5517.910156            NSW   

      gender  age merchant_suburb extraction_date extraction_time     amount  \
0          F   26         Ashmore      2018-08-01        01:01:15  16.250000   
1          F   26          Sydney      2018-08-01        01:13:45  14.190000   
2          M   38          Sydney      2018-08-01        01:26:15   6.420000   
3          F   40         Buderim      2018-08-01        01:38:45  40.900002   
4          F   26   Mermaid Beach      2018-08-01        01:51:15   3.250000   
...      ...  ...             ...             ...             ...        ...   
12038      F   30        Ringwood      2018-10-31        23:09:06   9.790000   
12039      M   20          Casula      2018-10-31        23:21:46  63.869999   
12040      M   28      Kings Park      2018-10-31        23:34:25  43.959999   
12041      M   69        Oakleigh      2018-10-31        23:47:05  30.770000   
12042      M   31          Mascot      2018-10-31        23:59:44  22.360001   

          customer_id merchant_long_lat movement  
0      CUS-2487424745     153.38 -27.99    debit  
1      CUS-2487424745     151.21 -33.87    debit  
2      CUS-2142601169     151.21 -33.87    debit  
3      CUS-1614226872     153.05 -26.68    debit  
4      CUS-2487424745     153.44 -28.06    debit  
...               ...               ...      ...  
12038    CUS-55310383     145.23 -37.81    debit  
12039  CUS-2688605418     150.88 -33.96    debit  
12040  CUS-2663907001     150.92 -33.74    debit  
12041  CUS-1388323263     145.09 -37.91    debit  
12042  CUS-3129499595     151.19 -33.93    debit  

[7717 rows x 16 columns]

# 

In [92]:
# Resetting the index
df2.reset_index(inplace = True)

In [93]:
df2

index      status  card_present_flag       long_lat txn_description  \
0         0  authorized                1.0  153.41 -27.95             POS   
1         1  authorized                0.0  153.41 -27.95       SALES-POS   
2         2  authorized                1.0  151.23 -33.94             POS   
3         3  authorized                1.0  153.10 -27.66       SALES-POS   
4         4  authorized                1.0  153.41 -27.95       SALES-POS   
...     ...         ...                ...            ...             ...   
7712  12038  authorized                0.0  149.83 -29.47             POS   
7713  12039  authorized                1.0  151.22 -33.87       SALES-POS   
7714  12040  authorized                1.0  151.12 -33.89             POS   
7715  12041  authorized                1.0  144.96 -37.76       SALES-POS   
7716  12042  authorized                1.0  150.92 -33.77       SALES-POS   

                               merchant_id       balance merchant_state  \
0     81c48296-73be-44a7-befa-d053f48ce7cd     35.389999            QLD   
1     830a451c-316e-4a6a-bf25-e37caedca49e     21.200001            NSW   
2     835c231d-8cdf-4e96-859d-e9d571760cf0      5.710000            NSW   
3     48514682-c78a-4a88-b0da-2d6302e64673   2117.219971            QLD   
4     b4e02c10-0852-4273-b8fd-7b3395e32eb0     17.950001            QLD   
...                                    ...           ...            ...   
7712  32aa73dc-b7c2-4161-b14d-6271b96ce792  14054.139648            VIC   
7713  296a0500-8552-48ac-ac81-ec37065b568e   9137.790039            NSW   
7714  e5975ab4-08f7-4725-a369-24cc0e35ed6e  45394.570312            NSW   
7715  af49051a-591d-4b08-bd3c-27730b70ed37  11350.669922            VIC   
7716  f31f4b14-2040-40ec-a120-b141bb274cbd   5517.910156            NSW   

     gender  age merchant_suburb extraction_date extraction_time     amount  \
0         F   26         Ashmore      2018-08-01        01:01:15  16.250000   
1         F   26          Sydney      2018-08-01        01:13:45  14.190000   
2         M   38          Sydney      2018-08-01        01:26:15   6.420000   
3         F   40         Buderim      2018-08-01        01:38:45  40.900002   
4         F   26   Mermaid Beach      2018-08-01        01:51:15   3.250000   
...     ...  ...             ...             ...             ...        ...   
7712      F   30        Ringwood      2018-10-31        23:09:06   9.790000   
7713      M   20          Casula      2018-10-31        23:21:46  63.869999   
7714      M   28      Kings Park      2018-10-31        23:34:25  43.959999   
7715      M   69        Oakleigh      2018-10-31        23:47:05  30.770000   
7716      M   31          Mascot      2018-10-31        23:59:44  22.360001   

         customer_id merchant_long_lat movement  
0     CUS-2487424745     153.38 -27.99    debit  
1     CUS-2487424745     151.21 -33.87    debit  
2     CUS-2142601169     151.21 -33.87    debit  
3     CUS-1614226872     153.05 -26.68    debit  
4     CUS-2487424745     153.44 -28.06    debit  
...              ...               ...      ...  
7712    CUS-55310383     145.23 -37.81    debit  
7713  CUS-2688605418     150.88 -33.96    debit  
7714  CUS-2663907001     150.92 -33.74    debit  
7715  CUS-1388323263     145.09 -37.91    debit  
7716  CUS-3129499595     151.19 -33.93    debit  

[7717 rows x 17 columns]

There are now two dataframes.

    df1 - contains all rows including NaN

    df2 - contains only rows with non NaN values

# Exploratory Data Analysis on df1

In [95]:
df1.head()

status  card_present_flag       long_lat txn_description  \
0  authorized                1.0  153.41 -27.95             POS   
1  authorized                0.0  153.41 -27.95       SALES-POS   
2  authorized                1.0  151.23 -33.94             POS   
3  authorized                1.0  153.10 -27.66       SALES-POS   
4  authorized                1.0  153.41 -27.95       SALES-POS   

                            merchant_id      balance merchant_state gender  \
0  81c48296-73be-44a7-befa-d053f48ce7cd    35.389999            QLD      F   
1  830a451c-316e-4a6a-bf25-e37caedca49e    21.200001            NSW      F   
2  835c231d-8cdf-4e96-859d-e9d571760cf0     5.710000            NSW      M   
3  48514682-c78a-4a88-b0da-2d6302e64673  2117.219971            QLD      F   
4  b4e02c10-0852-4273-b8fd-7b3395e32eb0    17.950001            QLD      F   

   age merchant_suburb extraction_date extraction_time     amount  \
0   26         Ashmore      2018-08-01        01:01:15  16.250000   
1   26          Sydney      2018-08-01        01:13:45  14.190000   
2   38          Sydney      2018-08-01        01:26:15   6.420000   
3   40         Buderim      2018-08-01        01:38:45  40.900002   
4   26   Mermaid Beach      2018-08-01        01:51:15   3.250000   

      customer_id merchant_long_lat movement  
0  CUS-2487424745     153.38 -27.99    debit  
1  CUS-2487424745     151.21 -33.87    debit  
2  CUS-2142601169     151.21 -33.87    debit  
3  CUS-1614226872     153.05 -26.68    debit  
4  CUS-2487424745     153.44 -28.06    debit

In [94]:
# Looking for average age
df1[['age','amount',
'balance']].mean()

age           30.582330
amount       187.933731
balance    14704.198242
dtype: float64

In [97]:
df.extraction_time.dt.weekday

AttributeError: Can only use .dt accessor with datetimelike values